### PR TITLE
Use new QCFractal

### PR DESCRIPTION
**Pull request by mattwthompson**

### CODE REVIEW

**Comment by mattwthompson** [comment]
Blocked in some capacity by a new QCSubmit

**Comment by mattwthompson** [comment]
Now blocked by QCFractal itself supporting v2

File: docs/getting-started/installation.md
Comment (mattwthompson) [inline docs/getting-started/installation.md:54]: #312 would make some of these older changes simpler

**Comment by mattwthompson** [comment]
Blocked by https://github.com/openforcefield/openff-qcsubmit/pull/232

File: docs/getting-started/installation.md
Comment (mattwthompson) [inline docs/getting-started/installation.md:75]: These changes I'm deciding to leave in place - the Python version is less important than the `dftd4-python` change, which I understand to be essential

Comment (j-wags) [reply docs/getting-started/installation.md:75]: Do you have a link for why this change is essential? My current (vague) understanding is that dftd3 and dftd4 have fundamentally different capabilities.

Comment (mattwthompson) [reply docs/getting-started/installation.md:75]: This is a dependency update in Psi4: https://github.com/openforcefield/openff-bespokefit/pull/308#issuecomment-1841921111

Comment (j-wags) [reply docs/getting-started/installation.md:75]: Ah, thanks for linking that context.

Haha I'm not sure this resolves it and I'm seeing we're having the same discussion as we had there.

I don't think `dftd4` is version 4 of the `dftd` package. I think `d3` and `d4` are fundamentally different things. 
BespokeFit's default level of theory is B3LYP-D3BJ/DZVP. I think the "D3" part of that alphabet soup maps to the `dftd3` package. 

Basically I think Lori is saying that back in the day, `psi4::dftd3` was packaged differently from `psi4::dftd4`. The latter was packaged more cleanly. Now, on `conda-forge`, both are packaged in a clean way. That is, `conda-forge::dftd4-python` and its deps are largely unchanged from their `psi4` channel versions. But `conda-forge::dftd3-python` and its deps are heavily changed from their `psi4` channel versions. 

Indeed, if I make a bespokefit env with the `dftd4-python` package (and not the dftd3 one), and then try to run a basic fit using the default level of theory:

```
openff-bespoke executor run --smiles 'CC' --workflow 'default' --target-torsion '[C:1]-[C:2]
```

I get an error that complains about missing dftd3

``` /Users/jeffreywagner/conda/envs/bespokefit-env/share/psi4/basis/dgauss-dzvp-mix.gbs \n    atoms 2, 5        entry C           
 line   192 file /Users/jeffreywagner/conda/envs/bespokefit-env/share/psi4/basis/dgauss-dzvp-mix.gbs \n\nTraceback (most recent
 call last):\n  File \"/Users/jeffreywagner/conda/envs/bespokefit-env/lib/python3.11/site-packages/qcengine/programs/base.py\",
 line 79, in get_program\n    ret.found(raise_error=True)\n  File                                                              
 \"/Users/jeffreywagner/conda/envs/bespokefit-env/lib/python3.11/site-packages/qcengine/programs/dftd_ng.py\", line 205, in    
 found\n    return which_import(\n           ^^^^^^^^^^^^^\n  File                                                             
 \"/Users/jeffreywagner/conda/envs/bespokefit-env/lib/python3.11/site-packages/qcelemental/util/importing.py\", line 45, in    
 which_import\n    raise ModuleNotFoundError(\nModuleNotFoundError: Python module 'dftd3' not found in envvar PYTHONPATH.      
 Please install a dftd3 version with enabled Python API (e.g. conda install dftd3-python -c conda-forge)\n\nThe above exception
 was the direct cause of the following exception:\n\nTraceback (most recent call last):\n  File                                
 \"/Users/jeffreywagner/conda/envs/bespokefit-env/lib/python3.11/site-packages/psi4/driver/schema_wrapper.py\", line 460, in   
 run_qcschema\n    ret_data = run_json_qcschema(input_model.dict(), clean, False, keep_wfn=keep_wfn)\n                         
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File                                                
 \"/Users/jeffreywagner/conda/envs/bespokefit-env/lib/python3.11/site-packages/psi4/driver/schema_wrapper.py\", line 626, in   
 run_json_qcschema\n    val, wfn = methods_dict_[json_data[\"driver\"]](method, **kwargs)\n                                    
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File                                                                  
 \"/Users/jeffreywagner/conda/envs/bespokefit-env/lib/python3.11/site-packages/psi4/driver/driver.py\", line 638, in gradient\n
 wfn = procedures['gradient'](lowername, molecule=molecule, **kwargs)\n                                                        
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File                                             
 \"/Users/jeffreywagner/conda/envs/bespokefit-env/lib/python3.11/site-packages/psi4/driver/procrouting/proc.py\", line 96, in  
 select_scf_gradient\n    return func(name, **kwargs)\n           ^^^^^^^^^^^^^^^^^^^^\n  File                                 
 \"/Users/jeffreywagner/conda/envs/bespokefit-env/lib/python3.11/site-packages/psi4/driver/procrouting/proc.py\", line 2688, in
 run_scf_gradient\n    ref_wfn = run_scf(name, **kwargs)\n              ^^^^^^^^^^^^^^^^^^^^^^^\n  File                        
 \"/Users/jeffreywagner/conda/envs/bespokefit-env/lib/python3.11/site-packages/psi4/driver/procrouting/proc.py\", line 2588, in
 run_scf\n    scf_wfn = scf_helper(name, post_scf=False, **kwargs)\n              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n 
 File \"/Users/jeffreywagner/conda/envs/bespokefit-env/lib/python3.11/site-packages/psi4/driver/procrouting/proc.py\", line    
 1851, in scf_helper\n    disp_energy = scf_wfn._disp_functor.compute_energy(scf_wfn.molecule(), scf_wfn)\n                    
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File                                                     
 \"/Users/jeffreywagner/conda/envs/bespokefit-env/lib/python3.11/site-packages/psi4/driver/procrouting/empirical_dispersion.py\
 ", line 264, in compute_energy\n    jobrec = qcng.compute(\n             ^^^^^^^^^^^^^\n  File                                
 \"/Users/jeffreywagner/conda/envs/bespokefit-env/lib/python3.11/site-packages/qcengine/compute.py\", line 79, in compute\n    
 executor = get_program(program)\n               ^^^^^^^^^^^^^^^^^^^^\n  File                                                  
 \"/Users/jeffreywagner/conda/envs/bespokefit-env/lib/python3.11/site-packages/qcengine/programs/base.py\", line 81, in        
 get_program\n    raise ResourceError(f\"Program {name} is registered with QCEngine, but cannot be found.\") from              
 err\nqcengine.exceptions.ResourceError: Program s-dftd3 is registered with QCEngine, but cannot be found.\n\n\n
```

however directly installing `dftd3-python` resolves this error.

Comment (mattwthompson) [reply docs/getting-started/installation.md:75]: I'm just going to remove all related changes as this is unrelated to the objective, they can be fixed elsewhere

File: docs/getting-started/installation.md
Comment (mattwthompson) [inline docs/getting-started/installation.md:69]: `python=3.10` should not be in docs that might get copy-pasted to the shell

**Review by j-wags — COMMENTED** [review]
An update since this is taking a little while: I did a once-over on the code and it looked fine. However I ran the quick-start calc 4 times using the old and new stack and am seeing consistent differences in the resulting parameters. So this post is just an update and I'm still looking into this. 

Here's where I've run them: [2024_03_25_quickstart_test.zip](https://github.com/openforcefield/openff-bespokefit/files/14763919/2024_03_25_quickstart_test.zip)

Here's my analysis result (copied from the notebook in that zip):

```python
from openff.toolkit import ForceField
import glob
folders = sorted(glob.glob("./*data*"))
ffs = dict()
for folder in folders:
    ffs[folder] = ForceField(f"{folder}/acetaminophen.offxml")
import numpy as np
from openff.units import unit
from matplotlib import pyplot
diffs = np.zeros((len(folders), len(folders))) * unit.kilocalorie_per_mole
for idx1, (folder1, ff1) in enumerate(ffs.items()):
    for idx2, (folder2, ff2) in enumerate(ffs.items()):
        tot_diff = 0
        for param1 in ff1["ProperTorsions"][-6:]:
            param2 = ff2["ProperTorsions"][param1.smirks]
            for k1, k2 in zip(param1.k, param2.k):
                tot_diff += abs(k1 - k2)
        #print(tot_diff)
        diffs[idx1, idx2] = tot_diff
pyplot.imshow(diffs)
pyplot.xticks(range(len(folders)), labels=folders, rotation=75)
pyplot.yticks(range(len(folders)), labels=folders)
pyplot.colorbar()
```
Screenshot 2024-03-26 at 12 45 33 PM

**Comment by j-wags** [comment]
Ahh, chalk it up to environment contamination. Turns out my new env had openeye and the old one didn't. Adding a run where I've removed OE from the new env now gives results that match the old results. 

Screenshot 2024-03-26 at 1 57 36 PM

File: openff/bespokefit/tests/conftest.py
Comment (j-wags) [inline openff/bespokefit/tests/conftest.py:336]: (not blocking) Not a huge deal since it's in a test file and the tests are doing fine, but I'm adding this in case it helps someone in the future. 

```suggestion
        properties=AtomicResultProperties( 
         # missing quadrupole moment, see 
         # https://github.com/openforcefield/openff-qcsubmit/pull/195/files#diff-38e27222b4aa42cac9e280107a45496a7f27ddf6d648c7dcf330ef5622f8499aR885 
         # for breadcrumbs if needed
```

File: openff/bespokefit/tests/schema/test_targets.py
Comment (j-wags) [inline openff/bespokefit/tests/schema/test_targets.py:145]: (not blocking) Could be good to leave a comment explaining this

**Review by j-wags — COMMENTED** [review]
I'm going to see if I can get codecov happy again before I approve.

**Comment by jthorton** [comment]
@j-wags thanks for doing that careful testing it seems like nothing is broken when running the full workflow! What does the `new_data_w_temp` refer to as the parameters seem to be slightly different to the `new_data` runs which seem very consistent?

File: openff/bespokefit/tests/schema/test_targets.py
Comment (mattwthompson) [reply openff/bespokefit/tests/schema/test_targets.py:145]: It's not used at all

File: openff/bespokefit/schema/data.py
Comment (mattwthompson) [reply openff/bespokefit/schema/data.py:130]: The corresponding method is not implemented, so the user will have a hard time hitting a but at all

**Review by j-wags — APPROVED** [review]
Codecov looks good. Merging!

**Comment by j-wags** [comment]
Oh, and @jthorton (sorry to miss your comment, the page didn't cleanly refresh when I opened this today) - `new_data_w_temp` is the new stack run with `BEFLOW_KEEP_TMP_FILES=True` so that I could start debugging using the intermediate files. Keeping the intermediate files shouldn't introduce any numerical differences in the fitting process so I don't think that the significant differentiator for that run. My guess is that it just hit some variability in charge assignment/the fitting process (as a breadcrumb: I did notice that in some runs the new parameters in the OFFXML are written in different orders - this may affect the order of iteration during FB fits and therefore the optimization trajectory?)

We see a little bit of the variation between the `old_data*/` runs as well using the legacy stack, and the `new_data_no_openeye` run matches a group of the `old...` outcomes closely.

I don't think this is a showstopper for release since some variability was always there. It would be better if this wasn't the case, but since this PR has been open 7 months now, I'd prefer to handle the variability problem in a separate issue/PR.

**Comment by j-wags** [comment]
@jthorton - Since you're currently listed as owner for BespokeFit, and since your previous comments mentioned other questions - Would you prefer those be addressed before a new release is cut? My team is over budget for BespokeFit effort this month (and behind on other goals) so we can't put in significant time until later in April. So either:

* We could cut the release now and handle other stuff in future releases
* We could hold off on the release and you could take the lead on investigating differences/adding tests
* We could hold off on the release and my team could could take the lead on the other questions in mid/late-April

**Comment by jthorton** [comment]
Hey @j-wags I think we should be fine to cut a release now as this update should not change the fitting outcomes and we can look into the differences in April, I'll be able to help with this but not take it on solo. 

Your graphs above got me thinking about the possible fitting regression reported in https://github.com/openforcefield/openff-bespokefit/issues/304 which is proving hard to track down which is the main thing I want to keep looking into.

**Pull request merged**
